File: python/nexus/services/resource_graph/connections.py
"""Hydrated connection reads over ``resource_edges``, scoped to one viewer."""

from __future__ import annotations

from collections import defaultdict
from datetime import datetime
from typing import Any, cast
from uuid import UUID

from sqlalchemy import and_, false, or_, select
from sqlalchemy.orm import Session

from nexus.db.models import NoteBlock, ResourceEdge
from nexus.errors import ApiErrorCode, InvalidRequestError
from nexus.services.resource_graph.citations import citation_reader_targets_for_edges
from nexus.services.resource_graph.edges import Pair, link_note_blocks_for_pairs
from nexus.services.resource_graph.refs import ResourceRef, ResourceScheme
from nexus.services.resource_graph.resolve import resolve_refs
from nexus.services.resource_graph.schemas import (
    Connection,
    ConnectionCitation,
    ConnectionEndpoint,
    ConnectionLinkNote,
    ConnectionPage,
    ConnectionQuery,
    EdgeKind,
    EdgeOrigin,
    is_neutral_link_shape,
    snapshot_from_jsonb,
)
from nexus.services.resource_items.capabilities import expand_owned_child_refs
from nexus.services.resource_items.routing import resource_activations_for_refs

_PREVIEW_CHARS = 200


def query_connections(db: Session, *, viewer_id: UUID, query: ConnectionQuery) -> ConnectionPage:
    if not query.refs:
        raise InvalidRequestError(ApiErrorCode.E_INVALID_REQUEST, "At least one ref is required")
    if len(query.refs) > 200:
        raise InvalidRequestError(ApiErrorCode.E_INVALID_REQUEST, "At most 200 refs are allowed")
    if query.limit < 1 or query.limit > 100:
        raise InvalidRequestError(ApiErrorCode.E_INVALID_REQUEST, "limit must be between 1 and 100")

    expanded: dict[str, ResourceRef] = {}
    for ref in query.refs:
        expanded.setdefault(ref.uri, ref)
        if query.rollup == "owner":
            for child in expand_owned_child_refs(db, viewer_id=viewer_id, ref=ref):
                expanded.setdefault(child.uri, child)
    matched: set[Pair] = {(ref.scheme, ref.id) for ref in expanded.values()}

    rows = _query_rows(db, viewer_id=viewer_id, refs=tuple(expanded.values()), query=query)
    page_rows = rows[: query.limit]
    next_cursor = (
        f"{page_rows[-1].created_at.isoformat()}|{page_rows[-1].id}"
        if len(rows) > query.limit
        else None
    )
    endpoints = _hydrate_endpoints(db, viewer_id=viewer_id, rows=page_rows)
    link_notes = _link_notes_for_rows(db, viewer_id=viewer_id, rows=page_rows)
    citations = citation_reader_targets_for_edges(
        db,
        viewer_id=viewer_id,
        edges=page_rows,
        target_missing_ref_uris={
            endpoint.ref.uri for endpoint in endpoints.values() if endpoint.missing
        },
        target_routeable_ref_uris={
            endpoint.ref.uri
            for endpoint in endpoints.values()
            if endpoint.activation.href is not None
        },
    )

    items: list[Connection] = []
    for row in page_rows:
        source_ref = ResourceRef(scheme=cast("ResourceScheme", row.source_scheme), id=row.source_id)
        target_ref = ResourceRef(scheme=cast("ResourceScheme", row.target_scheme), id=row.target_id)
        incoming = query.direction == "incoming" or (
            query.direction == "both"
            and (target_ref.scheme, target_ref.id) in matched
            and (source_ref.scheme, source_ref.id) not in matched
        )
        projection = citations.get(row.id)
        items.append(
            Connection(
                edge_id=row.id,
                # A neutral Link is undirected: no presenter may read meaning from its
                # canonical storage direction. ``other`` still points at the far endpoint.
                direction=(
                    "undirected"
                    if _is_neutral_link_row(row)
                    else ("incoming" if incoming else "outgoing")
                ),
                kind=cast("EdgeKind", row.kind),
                origin=cast("EdgeOrigin", row.origin),
                snapshot=snapshot_from_jsonb(row.snapshot) if row.snapshot is not None else None,
                source_order_key=row.source_order_key,
                target_order_key=row.target_order_key,
                ordinal=row.ordinal,
                source_ref=source_ref,
                target_ref=target_ref,
                source=endpoints[source_ref.uri],
                target=endpoints[target_ref.uri],
                other=endpoints[source_ref.uri if incoming else target_ref.uri],
                citation=(
                    ConnectionCitation(
                        ordinal=projection.ordinal,
                        role=projection.role,
                        snapshot=projection.snapshot,
                        activation=endpoints[target_ref.uri].activation,
                        target_media_id=projection.media_id,
                        target_locator=projection.locator,
                        target_status=projection.target_status,
                    )
                    if projection is not None
                    else None
                ),
                link_note=link_notes.get(row.id),
                created_at=row.created_at,
            )
        )
    return ConnectionPage(items=tuple(items), next_cursor=next_cursor)


def _query_rows(
    db: Session, *, viewer_id: UUID, refs: tuple[ResourceRef, ...], query: ConnectionQuery
) -> list[ResourceEdge]:
    directions: list[Any] = []
    if query.direction in ("incoming", "both"):
        directions.append(
            _endpoint_clause(ResourceEdge.target_scheme, ResourceEdge.target_id, refs)
        )
    if query.direction in ("outgoing", "both"):
        directions.append(
            _endpoint_clause(ResourceEdge.source_scheme, ResourceEdge.source_id, refs)
        )
    stmt = select(ResourceEdge).where(
        ResourceEdge.user_id == viewer_id,
        or_(*directions),
        # Structural Link-note attachment edges are folded onto their Link, never
        # rendered as their own connection.
        ResourceEdge.origin != "link_note",
    )
    if query.filters.origins is not None:
        stmt = stmt.where(ResourceEdge.origin.in_(query.filters.origins))
    if query.filters.kinds is not None:
        stmt = stmt.where(ResourceEdge.kind.in_(query.filters.kinds))
    if query.filters.source_schemes is not None:
        stmt = stmt.where(ResourceEdge.source_scheme.in_(query.filters.source_schemes))
    if query.filters.target_schemes is not None:
        stmt = stmt.where(ResourceEdge.target_scheme.in_(query.filters.target_schemes))
    if query.cursor is not None:
        created_at, edge_id = _decode_cursor(query.cursor)
        stmt = stmt.where(
            or_(
                ResourceEdge.created_at < created_at,
                and_(ResourceEdge.created_at == created_at, ResourceEdge.id < edge_id),
            )
        )
    return list(
        db.execute(
            stmt.order_by(ResourceEdge.created_at.desc(), ResourceEdge.id.desc()).limit(
                query.limit + 1
            )
        )
        .scalars()
        .all()
    )


def _endpoint_clause(scheme_column: Any, id_column: Any, refs: tuple[ResourceRef, ...]) -> Any:
    by_scheme: dict[ResourceScheme, list[UUID]] = defaultdict(list)
    for ref in refs:
        by_scheme[ref.scheme].append(ref.id)
    if not by_scheme:
        return false()
    return or_(
        *(and_(scheme_column == scheme, id_column.in_(ids)) for scheme, ids in by_scheme.items())
    )


def _hydrate_endpoints(
    db: Session, *, viewer_id: UUID, rows: list[ResourceEdge]
) -> dict[str, ConnectionEndpoint]:
    refs: dict[str, ResourceRef] = {}
    for row in rows:
        for scheme, resource_id in (
            (row.source_scheme, row.source_id),
            (row.target_scheme, row.target_id),
        ):
            refs.setdefault(
                f"{scheme}:{resource_id}",
                ResourceRef(scheme=cast("ResourceScheme", scheme), id=resource_id),
            )
    resolved = resolve_refs(db, viewer_id=viewer_id, refs=list(refs.values()))
    activations = resource_activations_for_refs(
        db,
        viewer_id=viewer_id,
        refs=list(refs.values()),
        missing_ref_uris={
            ref.uri for ref, item in zip(refs.values(), resolved, strict=True) if item.missing
        },
    )
    return {
        ref.uri: ConnectionEndpoint(
            ref=ref,
            label=item.label,
            description=item.summary or None,
            activation=activations[ref.uri],
            href=activations[ref.uri].href,
            missing=item.missing,
        )
        for ref, item in zip(refs.values(), resolved, strict=True)
    }


def _is_neutral_link_row(row: ResourceEdge) -> bool:
    return is_neutral_link_shape(
        origin=row.origin,
        kind=row.kind,
        ordinal=row.ordinal,
        snapshot=row.snapshot,
        source_order_key=row.source_order_key,
        target_order_key=row.target_order_key,
    )


def _link_notes_for_rows(
    db: Session, *, viewer_id: UUID, rows: list[ResourceEdge]
) -> dict[UUID, ConnectionLinkNote]:
    """Fold each neutral Link's note onto its edge; the attachment rows stay invisible."""
    pairs = {
        row.id: frozenset({(row.source_scheme, row.source_id), (row.target_scheme, row.target_id)})
        for row in rows
        if _is_neutral_link_row(row)
    }
    if not pairs:
        return {}
    notes = link_note_blocks_for_pairs(db, viewer_id=viewer_id, pairs=list(pairs.values()))
    if not notes:
        return {}
    previews = {
        block_id: (body_text[:_PREVIEW_CHARS] if body_text else None)
        for block_id, body_text in db.execute(
            select(NoteBlock.id, NoteBlock.body_text).where(
                NoteBlock.id.in_({note_id for ids in notes.values() for note_id in ids})
            )
        ).all()
    }
    return {
        edge_id: ConnectionLinkNote(
            ref=ResourceRef(scheme="note_block", id=notes[pair][0]),
            preview=previews.get(notes[pair][0]),
        )
        for edge_id, pair in pairs.items()
        if pair in notes
    }


def _decode_cursor(raw: str) -> tuple[datetime, UUID]:
    created_raw, separator, edge_id_raw = raw.partition("|")
    if not separator:
        raise InvalidRequestError(ApiErrorCode.E_INVALID_REQUEST, "Invalid cursor")
    try:
        return datetime.fromisoformat(created_raw), UUID(edge_id_raw)
    except ValueError as exc:
        raise InvalidRequestError(ApiErrorCode.E_INVALID_REQUEST, "Invalid cursor") from exc

File: apps/web/src/lib/resourceGraph/connections.ts
import type { ApiPath } from "@/lib/api/client";
import { apiFetch } from "@/lib/api/client";
import {
  parseResourceRef,
  RESOURCE_SCHEMES,
  type ResourceScheme,
} from "@/lib/resourceGraph/resourceRef";
import type { ResourceActivation } from "@/lib/resources/activation";
import {
  decodeResourceActionSubject,
  type ResourceActionSubject,
} from "@/lib/resources/resourceActionTarget";
import {
  expectArray,
  expectBoolean,
  expectExactRecord,
  expectInteger,
  expectNullableInteger,
  expectNullableString,
  expectOneOf,
  expectRecord,
  expectString,
} from "@/lib/validation";

// Edge vocabulary lives here (its natural home — connections is the sole
// remaining edge-shape-reading module). Mirrors
// `nexus/services/resource_graph/schemas.py`.
export const EDGE_KINDS = ["context", "supports", "contradicts"] as const;
export type EdgeKind = (typeof EDGE_KINDS)[number];

export const EDGE_ORIGINS = [
  "user",
  "citation",
  "system",
  "note_body",
  "highlight_note",
  "synapse",
  "document_embed",
  "assistant",
  "link_note",
] as const;
export type EdgeOrigin = (typeof EDGE_ORIGINS)[number];

export interface EdgeOut {
  id: string;
  kind: EdgeKind;
  origin: EdgeOrigin;
  source_ref: string;
  target_ref: string;
  source_order_key: string | null;
  target_order_key: string | null;
  ordinal: number | null;
  snapshot: Record<string, unknown> | null;
  source_label: string;
  source_missing: boolean;
  target_label: string;
  target_missing: boolean;
  created_at: string;
}

export interface ConnectionEndpointOut {
  ref: string;
  scheme: ResourceScheme;
  id: string;
  label: string | null;
  description: string | null;
  activation: ResourceActivation;
  href: string | null;
  missing: boolean;
}

export interface ConnectionActionEndpointOut extends ConnectionEndpointOut {
  actionSubject: ResourceActionSubject;
}

export interface ConnectionCitationOut {
  ordinal: number;
  role: EdgeKind;
  snapshot: Record<string, unknown>;
  activation: ResourceActivation;
  target_reader: ConnectionReaderTargetOut | null;
  target_status: "current" | "missing" | "forbidden" | "unanchorable";
}

export interface ConnectionReaderTargetOut {
  media_id: string | null;
  locator: Record<string, unknown> | null;
}

/**
 * The one ordinary note folded onto a user/context Link, resolved from its two
 * structural `link_note` attachment edges (which never surface as their own
 * connections). Distinct from `ConnectionCitationOut` — this is the Link's note,
 * not a citation projection.
 */
export interface ConnectionLinkNoteOut {
  ref: string;
  note_block_id: string;
  preview: string | null;
}

export interface ConnectionOut {
  edge_id: string;
  direction: "incoming" | "outgoing" | "undirected";
  kind: EdgeKind;
  origin: EdgeOrigin;
  snapshot: Record<string, unknown> | null;
  source_order_key: string | null;
  target_order_key: string | null;
  ordinal: number | null;
  source_ref: string;
  target_ref: string;
  source: ConnectionActionEndpointOut;
  target: ConnectionActionEndpointOut;
  other: ConnectionActionEndpointOut;
  citation: ConnectionCitationOut | null;
  link_note: ConnectionLinkNoteOut | null;
  created_at: string;
}

export interface ConnectionPage {
  items: ConnectionOut[];
  next_cursor: string | null;
}

export interface ConnectionSummaryOut {
  ref: string;
  total: number;
  by_kind: Record<string, number>;
  last_connected_at: string | null;
  dominant_kind: EdgeKind | null;
  top_peers: ConnectionEndpointOut[];
}

interface ConnectionSummaryResponse {
  data: { summaries: ConnectionSummaryOut[] };
}

/** Batch per-ref connection summaries (≤200 refs), AI/synapse excluded by default. */
export async function queryConnectionSummaries(
  refs: string[],
  options: { signal?: AbortSignal } = {},
): Promise<ConnectionSummaryOut[]> {
  if (refs.length === 0) {
    return [];
  }
  const response = await apiFetch<ConnectionSummaryResponse>(
    "/api/resource-graph/connections/summary" as ApiPath,
    {
      method: "POST",
      signal: options.signal,
      body: JSON.stringify({ refs }),
    },
  );
  return response.data.summaries;
}

export interface QueryConnectionsInput {
  refs: string[];
  direction: "incoming" | "outgoing" | "both";
  rollup?: "exact" | "owner";
  filters?: {
    origins?: EdgeOrigin[] | null;
    kinds?: EdgeKind[] | null;
    source_schemes?: ResourceScheme[] | null;
    target_schemes?: ResourceScheme[] | null;
  };
  limit?: number;
  cursor?: string | null;
}

interface QueryConnectionsResponse {
  data: unknown;
}

const CONNECTION_KEYS = [
  "edge_id",
  "direction",
  "kind",
  "origin",
  "snapshot",
  "source_order_key",
  "target_order_key",
  "ordinal",
  "source_ref",
  "target_ref",
  "source",
  "target",
  "other",
  "citation",
  "link_note",
  "created_at",
] as const;

function decodeNullableRecord(
  raw: unknown,
  name: string,
): Record<string, unknown> | null {
  return raw === null ? null : expectRecord(raw, name);
}

function decodeActionEndpoint(
  raw: unknown,
  name: string,
): ConnectionActionEndpointOut {
  const value = expectExactRecord(
    raw,
    [
      "ref",
      "scheme",
      "id",
      "label",
      "description",
      "activation",
      "href",
      "missing",
    ],
    name,
  );
  const ref = expectString(value.ref, `${name}.ref`);
  const parsedRef = parseResourceRef(ref);
  if (parsedRef === null) {
    // justify-defect: owned connection endpoints must identify a canonical
    // resource before they can enter the resource-action domain.
    throw new TypeError(`${name}.ref must be a canonical ResourceRef`);
  }
  const scheme = expectOneOf(value.scheme, RESOURCE_SCHEMES, `${name}.scheme`);
  const id = expectString(value.id, `${name}.id`);
  if (parsedRef.scheme !== scheme || parsedRef.id !== id) {
    // justify-defect: the endpoint's decomposed identity must describe the same
    // resource as its canonical ref.
    throw new TypeError(`${name}.scheme/id must equal ${name}.ref`);
  }
  const missing = expectBoolean(value.missing, `${name}.missing`);
  const activation = decodeConnectionActivation(
    value.activation,
    `${name}.activation`,
  );
  const actionSubject = decodeResourceActionSubject(
    { ref },
    `${name}.actionSubject`,
  );
  if (activation.resourceRef !== actionSubject.ref) {
    // justify-defect: occurrence navigation and canonical action identity are
    // separate facts, but an endpoint must publish both for the same resource.
    throw new TypeError(
      `${name}.activation.resource_ref must equal ${name}.ref`,
    );
  }
  const href = expectNullableString(value.href, `${name}.href`);
  if (href !== activation.href) {
    // justify-defect: the endpoint cannot publish two different activation
    // destinations for the same resource.
    throw new TypeError(`${name}.href must equal ${name}.activation.href`);
  }
  return {
    ref,
    scheme,
    id,
    label: expectNullableString(value.label, `${name}.label`),
    description: expectNullableString(value.description, `${name}.description`),
    activation,
    href,
    missing,
    actionSubject,
  };
}

function decodeConnectionActivation(
  raw: unknown,
  name: string,
): ResourceActivation {
  const value = expectExactRecord(
    raw,
    ["resource_ref", "kind", "href", "unresolved_reason"],
    name,
  );
  const ref = expectString(value.resource_ref, `${name}.resource_ref`);
  const activation = {
    resourceRef: ref,
    kind: expectOneOf(
      value.kind,
      ["route", "external", "none"] as const,
      `${name}.kind`,
    ),
    href: expectNullableString(value.href, `${name}.href`),
    unresolvedReason: expectNullableString(
      value.unresolved_reason,
      `${name}.unresolved_reason`,
    ),
  };
  decodeResourceActionSubject({ ref }, `${name}.subject`);
  return activation;
}

function decodeConnectionReaderTarget(
  raw: unknown,
  name: string,
): ConnectionReaderTargetOut {
  const value = expectExactRecord(raw, ["media_id", "locator"], name);
  return {
    media_id: expectNullableString(value.media_id, `${name}.media_id`),
    locator: decodeNullableRecord(value.locator, `${name}.locator`),
  };
}

function decodeConnectionCitation(
  raw: unknown,
  name: string,
): ConnectionCitationOut | null {
  if (raw === null) return null;
  const value = expectExactRecord(
    raw,
    [
      "ordinal",
      "role",
      "snapshot",
      "activation",
      "target_reader",
      "target_status",
    ],
    name,
  );
  return {
    ordinal: expectInteger(value.ordinal, `${name}.ordinal`),
    role: expectOneOf(value.role, EDGE_KINDS, `${name}.role`),
    snapshot: expectRecord(value.snapshot, `${name}.snapshot`),
    activation: decodeConnectionActivation(
      value.activation,
      `${name}.activation`,
    ),
    target_reader:
      value.target_reader === null
        ? null
        : decodeConnectionReaderTarget(
            value.target_reader,
            `${name}.target_reader`,
          ),
    target_status: expectOneOf(
      value.target_status,
      ["current", "missing", "forbidden", "unanchorable"] as const,
      `${name}.target_status`,
    ),
  };
}

function decodeConnectionLinkNote(
  raw: unknown,
  name: string,
): ConnectionLinkNoteOut | null {
  if (raw === null) return null;
  const value = expectExactRecord(
    raw,
    ["ref", "note_block_id", "preview"],
    name,
  );
  const ref = expectString(value.ref, `${name}.ref`);
  const parsedRef = parseResourceRef(ref);
  const noteBlockId = expectString(
    value.note_block_id,
    `${name}.note_block_id`,
  );
  if (parsedRef?.scheme !== "note_block" || parsedRef.id !== noteBlockId) {
    // justify-defect: a folded Link note must identify exactly one canonical
    // note block across both of its wire identity fields.
    throw new TypeError(`${name}.ref must identify ${name}.note_block_id`);
  }
  return {
    ref,
    note_block_id: noteBlockId,
    preview: expectNullableString(value.preview, `${name}.preview`),
  };
}

export function decodeConnectionOut(
  raw: unknown,
  name = "Connection",
): ConnectionOut {
  const value = expectExactRecord(raw, CONNECTION_KEYS, name);
  const sourceRef = expectString(value.source_ref, `${name}.source_ref`);
  const targetRef = expectString(value.target_ref, `${name}.target_ref`);
  if (
    parseResourceRef(sourceRef) === null ||
    parseResourceRef(targetRef) === null
  ) {
    // justify-defect: connection edges cannot enter the frontend with
    // non-canonical endpoint refs.
    throw new TypeError(`${name} source/target refs must be canonical`);
  }
  const source = decodeActionEndpoint(value.source, `${name}.source`);
  const target = decodeActionEndpoint(value.target, `${name}.target`);
  const other = decodeActionEndpoint(value.other, `${name}.other`);
  if (
    source.ref !== sourceRef ||
    target.ref !== targetRef ||
    (other.ref !== sourceRef && other.ref !== targetRef)
  ) {
    // justify-defect: hydrated connection endpoints must agree with the edge
    // identity fields that own the relationship.
    throw new TypeError(`${name} endpoint refs must match the edge refs`);
  }
  return {
    edge_id: expectString(value.edge_id, `${name}.edge_id`),
    direction: expectOneOf(
      value.direction,
      ["incoming", "outgoing", "undirected"] as const,
      `${name}.direction`,
    ),
    kind: expectOneOf(value.kind, EDGE_KINDS, `${name}.kind`),
    origin: expectOneOf(value.origin, EDGE_ORIGINS, `${name}.origin`),
    snapshot: decodeNullableRecord(value.snapshot, `${name}.snapshot`),
    source_order_key: expectNullableString(
      value.source_order_key,
      `${name}.source_order_key`,
    ),
    target_order_key: expectNullableString(
      value.target_order_key,
      `${name}.target_order_key`,
    ),
    ordinal: expectNullableInteger(value.ordinal, `${name}.ordinal`),
    source_ref: sourceRef,
    target_ref: targetRef,
    source,
    target,
    other,
    citation: decodeConnectionCitation(value.citation, `${name}.citation`),
    link_note: decodeConnectionLinkNote(value.link_note, `${name}.link_note`),
    created_at: expectString(value.created_at, `${name}.created_at`),
  };
}

function decodeConnectionPage(raw: unknown): ConnectionPage {
  const value = expectExactRecord(
    raw,
    ["items", "next_cursor"],
    "ConnectionPage",
  );
  return {
    items: expectArray(
      value.items,
      (item, index) =>
        decodeConnectionOut(item, `ConnectionPage.items[${index}]`),
      "ConnectionPage.items",
    ),
    next_cursor: expectNullableString(
      value.next_cursor,
      "ConnectionPage.next_cursor",
    ),
  };
}

export async function queryConnections(
  input: QueryConnectionsInput,
  options: { signal?: AbortSignal } = {},
): Promise<ConnectionPage> {
  const response = await apiFetch<QueryConnectionsResponse>(
    "/api/resource-graph/connections/query" as ApiPath,
    {
      method: "POST",
      signal: options.signal,
      body: JSON.stringify(input),
    },
  );
  return decodeConnectionPage(response.data);
}
